### PR TITLE
[CARBONDATA-1866] refactored CarbonLateDecodeRule to split different …

### PR DESCRIPTION
**Pull request by rahulforallp**
…rule for better usablity

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed? **No**
 
 - [X] Any backward compatibility impacted? **No**
 
 - [X] Document update required? **No**

 - [X] Testing done
        This PR contains refactored code of **CarbonLateDecodeRule** . So no new test-case required.
       Only UT, FT and SDV test success report required from CI.

 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/498/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/500/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1761/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1763/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2144/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/520/

**Comment by rahulforallp** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/521/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2153/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1781/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1788/

File: integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.mutation.CarbonProjectForUpdateCommand
+
+/**
+ * Rule specific for IUD operations
+ */
+class CarbonIUDRule extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (CarbonDecodeRuleHelper.checkIfRuleNeedToBeApplied(plan)) {
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala]: This check is not required, as IUD node once replaced will not present

File: integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonPlanTransformRule.scala [hunk 2]
```diff
@@ -44,37 +44,14 @@ import org.apache.carbondata.spark.CarbonAliasDecoderRelation
  * 1. Change the datatype for dictionary encoded column
  * 2. Add the dictionary decoder operator at appropriate place.
  */
-class CarbonLateDecodeRule extends Rule[LogicalPlan] with PredicateHelper {
+class CarbonPlanTransformRule extends Rule[LogicalPlan] with PredicateHelper {
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonPlanTransformRule.scala]: This is still lateDecodeRule only, not be CarbonPlanTransformRule

File: integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonUDFTransformRule.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.types.StringType
+
+import org.apache.carbondata.core.constants.CarbonCommonConstants
+
+class CarbonUDFTransformRule extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+
+    if (CarbonDecodeRuleHelper.checkIfRuleNeedToBeApplied(plan)) {
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonUDFTransformRule.scala]: remove this check

File: integration/spark2/src/main/spark2.2/CarbonSessionState.scala [hunk 4]
```diff
@@ -176,31 +178,40 @@ class CarbonSessionStateBuilder(sparkSession: SparkSession,
...
-      override val postHocResolutionRules: Seq[Rule[LogicalPlan]] =
-        new DetermineTableStats(session) +:
-        RelationConversions(conf, catalog) +:
-        PreprocessTableCreation(session) +:
-        PreprocessTableInsertion(conf) +:
-        DataSourceAnalysis(conf) +:
-        HiveAnalysis +:
-        customPostHocResolutionRules
-    }
+      new Analyzer(catalog, conf) {
```
Comment (gvramana) [inline integration/spark2/src/main/spark2.2/CarbonSessionState.scala]: Use customResolutionRules , not required to add extra interface

File: integration/spark2/src/main/spark2.2/CarbonSessionState.scala [hunk 5]
```diff
@@ -213,6 +224,13 @@ class CarbonOptimizer(
   extends SparkOptimizer(catalog, conf, experimentalMethods) {
 
   override def execute(plan: LogicalPlan): LogicalPlan = {
+    val transFormedPlan: LogicalPlan = CarbonOptimizerUtil.transformForScalarSubQuery(plan)
```
Comment (gvramana) [inline integration/spark2/src/main/spark2.2/CarbonSessionState.scala]: This changes not required

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/576/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1820/

**Comment by rahulforallp** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/598/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2197/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1828/

**Comment by gvramana** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/601/

**Pull request closed without merging**